---
title: "RNASeq downstream"
author: "Ankitha"
date: "2024-12-05"
output: html_document
---
#Install packages
```{r, echo=FALSE,message=FALSE,warning=FALSE}
#install pasilla package
if (!require("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install("pasilla")
if (!requireNamespace("AnnotationDbi", quietly = TRUE)) install.packages("AnnotationDbi")
if (!requireNamespace("rio", quietly = TRUE)) install.packages("rio")
if (!requireNamespace("org.Mm.eg.db", quietly = TRUE)) install.packages("org.Mm.eg.db")
if (!requireNamespace("clusterProfiler", quietly = TRUE)) BiocManager::install("clusterProfiler")
```

#Load libraries
```{r, echo=FALSE,message=FALSE,warning=FALSE}
library("DESeq2")
library("ggplot2")
library("ggpubr")
library("rio")
library("dplyr")
library("clusterProfiler")
library("org.Mm.eg.db")
library("AnnotationDbi")
library("ggrepel")
library("grid")
library("rio")
library(DOSE)
library(enrichplot)
library(EnhancedVolcano)
```

#Load functions
```{r, echo=FALSE,message=FALSE,warning=FALSE}
#Perform the pairwise analysis, print number of differently expressed genes and convert result into a data frame
run_pairwise_analysis <- function(contrast,comparison_name,padj_threshold=0.05){
  #Perform the comparison between the groups specified in the contrast.
  res <- results(dds, contrast = contrast)
  
  # Generate MA plot
  plotMA(res, main = paste("MA Plot -", comparison_name))
  
  #converting the result into a dataframe and adding the gene name, by calling the gene name function
  result_dataframe <- add_gene_names(res)
  
  #Number of differently expressed gene
  print(comparison_name)
  num_de_genes <- sum(res$padj < padj_threshold, na.rm = TRUE)
  print(paste("Total number of differentially expressed genes:", num_de_genes))
  print(summary(res, alpha = padj_threshold))
  
  return(result_dataframe)
}

#retrieving the gene name from the ensembl ID
add_gene_names <- function(res) {
  gene_names <- mapIds(
    org.Mm.eg.db,
    keys = rownames(res),
    column = "SYMBOL",
    keytype = "ENSEMBL",
    multiVals = "first"
  )
  # Adding the column GeneName
  res <- as.data.frame(res)
  res$GeneName <- gene_names
  return(res)
}

#Yolcano plot
generating_volcano_plot <- function(DE_result,comparison_name,padj_threshold=0.05,logfc_threshold=2){
  # Adding the column expression to the dataframe
  DE_result <- DE_result %>%
    mutate(
      Expression = case_when(
        log2FoldChange >= logfc_threshold & padj <= padj_threshold ~ "Up-regulated",
        log2FoldChange <= -logfc_threshold & padj <= padj_threshold ~ "Down-regulated",
        TRUE ~ "Not significant"
      )
    )

  # Generate the volcano_plot
  volcano_plot <- ggplot(DE_result, aes(log2FoldChange, -log(pvalue, 10))) + # -log10 conversion
    geom_point(aes(color = Expression), size = 2, alpha = 0.5) +
    xlab(expression("log"[2]*"FC")) +
    ylab(expression("-log"[10]*"P-Value")) +
    scale_color_manual(values = c("#043BE3", "grey", "#F7260F")) +
    xlim(-10, 10) +
    ggtitle(paste("Volcano Plot -", comparison_name)) +
    theme_classic() +
    theme(axis.text.x= element_text(colour = "black", size = 12),
              axis.text.y= element_text(colour = "black", size = 12),
              legend.text = element_text(colour = "black", size = 10),
              legend.title = element_text(colour = "black", size = 12, face = "bold"))
  
  print(volcano_plot)
}

#Print the top 10 most upregulated and downregulated genes, with their gene names
print_top_10_genes <- function(DE_result, comparison_name, logfc_threshold=2) {
  # Filter for genes with significant fold change (up-regulated and down-regulated)
  upregulated_genes <- DE_result[DE_result$log2FoldChange >= logfc_threshold & DE_result$padj <= 0.05, ]
  downregulated_genes <- DE_result[DE_result$log2FoldChange <= -logfc_threshold & DE_result$padj <= 0.05, ]
  
  # Get the top 10 most upregulated genes
  top_upregulated <- upregulated_genes[order(upregulated_genes$log2FoldChange, decreasing = TRUE), ]
  top_upregulated <- head(top_upregulated, 10)
  
  # Get the top 10 most downregulated genes 
  top_downregulated <- downregulated_genes[order(downregulated_genes$log2FoldChange), ]
  top_downregulated <- head(top_downregulated, 10)
  
  # Print the results
  print(paste("Top 10 upregulated genes -", comparison_name))
  print(top_upregulated[, c("GeneName", "log2FoldChange", "padj")])
  
  print(paste("Top 10 downregulated genes -", comparison_name))
  print(top_downregulated[, c("GeneName", "log2FoldChange", "padj")])
}

#GO analysis
run_go_analysis_pairwise <- function(DE_result,comparaison_name,ont,regulation="all",show_Category=10){
  
  ontology_list = list("BP"="Biological Processes","CC"="Cellular Component","MF"="Molecular Function")
  ontology = ontology_list[ont] 
  
  # Filter based on regulation type: "up", "down", or "all"
  if (regulation == "up") {
    # Upregulated genes: log2FoldChange >= 1 (1) and padj <= 0.05
    genes_de <- rownames(DE_result[DE_result$log2FoldChange >= 2 & DE_result$padj <= 0.05, ])
  } else if (regulation == "down") {
    # Downregulated genes: log2FoldChange <= -1 (1) and padj <= 0.05
    genes_de <- rownames(DE_result[DE_result$log2FoldChange <= -2 & DE_result$padj <= 0.05, ])
  } else {
    # All genes with padj <= 0.05
    genes_de <- rownames(DE_result[DE_result$padj <= 0.05 & !is.na(DE_result$padj), ])
  }
  
  # List of all genes measured in the analysis
  genes_universe <- rownames(DE_result)
  
  # Perform the GO analysis using enrichGO
  go_results <- enrichGO(
    gene = genes_de,
    universe = genes_universe,
    OrgDb = org.Mm.eg.db,
    ont = ont,
    keyType = "ENSEMBL",
    pAdjustMethod = "BH", #Adjust p-values using the Benjamini-Hochberg method
    pvalueCutoff = 0.05,
    qvalueCutoff = 0.05 #accept that 5% of tests found to be stats signif will be FP, FDR
  )
  
  #get simplified terms
  simplified_ego <- simplify(go_results)

  #set gradient colours for plots
  options(enrichplot.colours = c("red","blue"))
  
  # Create a dotplot for GO terms
  title= paste(ontology)
  dot_plot <- dotplot(simplified_ego, title =title,font.size = 20,showCategory =show_Category, color = "p.adjust")+
    theme_bw()
  print(dot_plot)
}
```
#Gene names
```{r, echo=FALSE,message=FALSE,warning=FALSE}

genes_article <- c("Unc93b1", "Slamf7", "Ms4a6b", "Ms4a6c", "Trim30a", "Pml", 
           "Ms4a4c", "Sp100", "Oas1g", "Mx1", "Oas3", "Zbp1", "Oas2", 
           "Mb21d1", "Phf11d", "Dhx58", "Phf11a", "Irf9", "Lgals9", 
           "Usp18", "Lair1", "Irf7", "Phf11b", "Xaf1", "Eif2ak2", 
           "Rsad2", "Gm4955", "Fcgr1", "Ifit3", "Siglec1", "Aif1", 
           "Ifit1", "Oas1a", "Oasl2", "Rtp4", "Gm5431", "Rnf213", 
           "Ifih1", "Oasl1", "Apobec1", "Lst1", "Adap1", "Grn", 
           "Epsti1", "Ddx60", "Apol9a", "Trim30d", "Dck")
```

#Plot aesthetics
```{r, echo=FALSE,message=FALSE,warning=FALSE}
plot_themes <- theme(axis.text.x= element_text(colour = "black", size = 12),
              axis.text.y= element_text(colour = "black", size = 12),
              axis.title=element_text(colour = "black", size=14),
              plot.title = element_text(colour = "black", face="bold"),
              legend.text = element_text(colour = "black", size = 10),
              legend.title = element_text(colour = "black", size = 12, face = "bold"))
```

#Load and clean data
```{r, echo=FALSE,message=FALSE,warning=FALSE}
#load data, row.names: sets the first col as row names 
data_count <- read.delim("gene_count.txt", header = T, row.names = 1)

#clean column names to just have the sample names
colnames(data_count) <- gsub(".*mapping\\.|sorted.bam.*", "", colnames(data_count))

#remove chr, start, end, strand, and length columns
data_count <- data_count[ -c(1:5) ]

#load group data
data_col <- read.csv("Samples.csv", header = T, row.names = 1)
data_col$Tissue <- as.factor(data_col$Tissue)
data_col$Treatment <- as.factor(data_col$Treatment) #sets groups as categorical variables

#check if rows from group data and columns from the feature counts are the same
#it is important that the rows in group data and columns in count matrix refer to the same groups, as DESeq2 does not make guesses
all(rownames(data_col) == colnames(data_count))
```

#Create DESeq dataset
```{r, echo=FALSE,message=FALSE,warning=FALSE}
#load group data
data_col <- read.csv("Groups.csv", header = T, row.names = 1)
data_col$Group <- as.factor(data_col$Group) #sets groups as categorical variables

#check if rows and col are the same
all(rownames(data_col) == colnames(data_count))

#data frame for anaylsis 
dds <- DESeqDataSetFromMatrix(countData = data_count,
                              colData = data_col,
                              design = ~ Group)

result_DEseq <- DESeq(dds)

#remove dependence of variance on mean, to stabilise the variance across gene counts
#This ensure more accurate results
result_DEseq_vst <- vst(result_DEseq, blind = TRUE)
```

#PCA plot
```{r, echo=FALSE,message=FALSE,warning=FALSE}
plotPCA(result_DEseq_vst, intgroup = "Group") +
  theme_bw() +
  scale_color_manual(name = "Groups",
                     labels = c("Blood WT Infected", "Blood WT Uninfected", "Lung WT Infected", "Lung WT Uninfected"), 
                     values = c("#DE2820", "#E57259", "#2062DE", "#7CA2E9"))

```

#Pairwise comparisons
#Lung case vs control
```{r, echo=FALSE,message=FALSE,warning=FALSE}
#Lung case vs control 
comparaison_name = "Lung case v control"
dds <- DESeq(dds)
#MA plot
df_lung_case_control <- run_pairwise_analysis(c("Group", "Lung_WT_Case", "Lung_WT_Control"), comparaison_name, padj_threshold=0.05)

#export dataframe to excel
tmp_file <- "/Users/ankithak/Documents/04 Bern/01 UniBe/06 RNA seq/rnaseq_downstream/DE_lung_forgsea.xlsx"
df <- df_lung_case_control %>%
  tibble::rownames_to_column(var = "Gene_ID")
export(df, tmp_file)
```

##Differential expression analysis
```{r, echo=FALSE,message=FALSE,warning=FALSE}
#volcano plot, only genes that are significantly differentially expressed are coloured 
volcano_plot <- generating_volcano_plot(df_lung_case_control,comparaison_name,padj_threshold=0.05,logfc_threshold=2) 

#Sort by significance and select top 10 upregulated genes
top_up_genes <- df_lung_case_control[order(df_lung_case_control$padj), ][1:10, ]

#Filter for significantly downregulated genes (e.g., padj < 0.05 if applicable)
downregulated_genes <- df_lung_case_control[df_lung_case_control$log2FoldChange < 0, ]
# Order by padj and then select the top 10 most downregulated genes
top_down_genes <- downregulated_genes[order(downregulated_genes$padj, downregulated_genes$log2FoldChange), ][1:10, ]

#Add labels for the top 10 upregulated genes
p <- volcano_plot + geom_text_repel(
    data = top_up_genes,
    aes(label = GeneName),
    box.padding = 0.5,
    max.overlaps = Inf,
    size = 3,
    color = "black"
  )

#Add labels for the top 10 downregulated genes
p <- p + geom_text_repel(
    data = top_down_genes,
    aes(label = GeneName),
    box.padding = 0.5,
    max.overlaps = Inf,
    size = 3,
    color = "black"
  )

#Add lines to differentiate between significiantly DE or not 
p + geom_vline(xintercept=c(-2, 2), col="black", linetype="dashed") +
   geom_hline(yintercept=-log10(0.05), col="black", linetype="dashed")

#gene names 
filter_comparison = df_lung_case_control[df_lung_case_control$GeneName %in% genes_article,c("log2FoldChange","pvalue","GeneName")]
print(filter_comparison)

#top 10 genes
print_top_10_genes(df_lung_case_control,comparaison_name)
```

##Overrepresnation analysis
```{r, echo=FALSE,message=FALSE,warning=FALSE}
#GO analysis
#All
#Biological process
BP_all <- run_go_analysis_pairwise(df_lung_case_control,comparaison_name,ont="BP",regulation = "all")
p1 <- BP_all + plot_themes
#Molecular function
MF_all <- run_go_analysis_pairwise(df_lung_case_control,comparaison_name,ont="MF",regulation = "all")
p2 <- MF_all + plot_themes
#Cellular component
CC_all <- run_go_analysis_pairwise(df_lung_case_control,comparaison_name,ont="CC",regulation = "all")
p3 <- CC_all + plot_themes

ggarrange(p1, p2, p3, ncol = 3, nrow = 1)

```

##Gene set enrichment analysis
```{r, echo=FALSE,message=FALSE,warning=FALSE}
d <- read.csv("DE_lung_forgsea.csv")

## feature 1: numeric vector
geneList = d[,2]
## feature 2: named vector
names(geneList) = as.character(d[,1])
## feature 3: decreasing order
geneList = sort(geneList, decreasing = TRUE)

#GSSizes were modified based on ontology type, as structure and size of gene sets may differ betweent these GOs.

#Biological process
ego_bp <- gseGO(geneList     = geneList,
              OrgDb        = org.Mm.eg.db,
              ont          = "BP",
              keyType = "ENSEMBL",
              minGSSize    = 100,
              maxGSSize    = 700,
              pvalueCutoff = 0.05,
              verbose      = FALSE)
#Molecular function
ego_mf <- gseGO(geneList     = geneList,
              OrgDb        = org.Mm.eg.db,
              ont          = "MF",
              keyType = "ENSEMBL",
              minGSSize    = 50,
              maxGSSize    = 300,
              pvalueCutoff = 0.05,
              verbose      = FALSE)
#Ceulluar component
ego_cc <- gseGO(geneList     = geneList,
              OrgDb        = org.Mm.eg.db,
              ont          = "CC",
              keyType = "ENSEMBL",
              minGSSize    = 25,
              maxGSSize    = 250,
              pvalueCutoff = 0.05,
              verbose      = FALSE)

#Visualisation
gset_BP <- dotplot(ego_bp, showCategory = 10, title = "Biological process")
p1 <- gset_BP + plot_themes
gset_MF <- dotplot(ego_mf, showCategory = 10, title = "Molecular function")
p2 <- gset_MF + plot_themes
gset_CC <- dotplot(ego_cc, showCategory = 10, title = "Cellular component")
p3 <- gset_CC + plot_themes

ggarrange(p1, p2, p3, ncol = 3, nrow = 1)

```

